scertecd/scertecd.go

```diff
@@ -1125,3 +1175,45 @@ func (s *Server) getOCSPResponse(ctx context.Context, leaf, issuer *x509.Certifi
 	}
 	return res, nil
 }
+
+// checkAWSPermissionsLoop is a background goroutine that periodically checks
+// whether our Route53 IAM permissions are still valid. This is meant to protect
+// us from moving the certd server between VMs and not having the right roles on
+// the new VM and then not noticing until certs fail to expire.
+//
+// On failure, this sets a metric on the server that we can then alert.
+func (s *Server) checkAWSPermissionsLoop() {
+	for {
+		if err := s.checkAWSPermissions(); err != nil {
+			s.Logf("checkAWSPermissions error: %v", err)
+			s.addError(errTypeMakeRecord)
+		}
+		time.Sleep(1 * time.Hour)
+	}
+}
+
+// checkAWSPermissions makes a test Route53 record to see if we have
+// suitable permissions.
+func (s *Server) checkAWSPermissions() error {
+	ctx, cancel := context.WithTimeout(context.Background(), 2*time.Minute)
+	defer cancel()
+
+	var buf bytes.Buffer
+	lg := log.New(&buf, "", log.LstdFlags|log.Lmsgprefix)
+	logf := lg.Printf
+
+	domain := s.dts[0].domain // TODO(bradfitz): random? check each unique domain?
+	release, err := s.acquireDomainRenewalLock(ctx, logf, domain)
+	if err != nil {
+		return err
```

Comment:
Should we include the contents of buf here too?

Reply:
There shouldn't be anything interesting in it at this point: the only way it could've failed is a timeout and it'll already be logged by the caller. Repeating it would probably just be stutter?

Reply:
I only noted it because we passed in the buffering logger.

Reply:
That logger is primarily to pass to makeRecord. In the happy path I don't want to see its logging when it should work. But if it fails, I want to see details.

But then once I get that logf func, I just passed it to acquireDomainRenewalLock too rather than making it spam or ioutil.Discard as well.